model/src/gateway/event/mod.rs

```diff
@@ -349,3 +349,68 @@ impl Display for EventConversionError {
+    const_assert_eq!(32, mem::size_of::<Connecting>());
+    const_assert_eq!(40, mem::size_of::<Disconnected>());
+    const_assert_eq!(632, mem::size_of::<GuildCreate>());
+    const_assert_eq!(16, mem::size_of::<GuildDelete>());
+    const_assert_eq!(32, mem::size_of::<GuildEmojisUpdate>());
+    const_assert_eq!(8, mem::size_of::<GuildIntegrationsUpdate>());
+    const_assert_eq!(400, mem::size_of::<GuildUpdate>());
+    const_assert_eq!(16, mem::size_of::<Identifying>());
+    const_assert_eq!(632, mem::size_of::<IntegrationCreate>());
+    const_assert_eq!(24, mem::size_of::<IntegrationDelete>());
+    const_assert_eq!(632, mem::size_of::<IntegrationUpdate>());
+    const_assert_eq!(16, mem::size_of::<InteractionCreate>());
+    const_assert_eq!(304, mem::size_of::<InviteCreate>());
+    const_assert_eq!(40, mem::size_of::<InviteDelete>());
+    const_assert_eq!(288, mem::size_of::<MemberAdd>());
+    const_assert_eq!(184, mem::size_of::<MemberRemove>());
+    const_assert_eq!(288, mem::size_of::<MemberUpdate>());
+    const_assert_eq!(112, mem::size_of::<MemberChunk>());
+    const_assert_eq!(1872, mem::size_of::<MessageCreate>());
+    const_assert_eq!(24, mem::size_of::<MessageDelete>());
+    const_assert_eq!(40, mem::size_of::<MessageDeleteBulk>());
+    const_assert_eq!(352, mem::size_of::<MessageUpdate>());
+    const_assert_eq!(24, mem::size_of::<Payload>());
+    const_assert_eq!(736, mem::size_of::<PresenceUpdate>());
+    const_assert_eq!(360, mem::size_of::<ReactionAdd>());
+    const_assert_eq!(360, mem::size_of::<ReactionRemove>());
+    const_assert_eq!(24, mem::size_of::<ReactionRemoveAll>());
+    const_assert_eq!(64, mem::size_of::<ReactionRemoveEmoji>());
+    const_assert_eq!(264, mem::size_of::<Ready>());
+    const_assert_eq!(8, mem::size_of::<Reconnecting>());
+    const_assert_eq!(16, mem::size_of::<Resuming>());
+    const_assert_eq!(128, mem::size_of::<RoleCreate>());
+    const_assert_eq!(16, mem::size_of::<RoleDelete>());
+    const_assert_eq!(128, mem::size_of::<RoleUpdate>());
+    const_assert_eq!(56, mem::size_of::<StageInstanceCreate>());
+    const_assert_eq!(56, mem::size_of::<StageInstanceDelete>());
+    const_assert_eq!(56, mem::size_of::<StageInstanceUpdate>());
+    const_assert_eq!(712, mem::size_of::<ThreadCreate>());
+    const_assert_eq!(32, mem::size_of::<ThreadDelete>());
+    const_assert_eq!(80, mem::size_of::<ThreadListSync>());
+    const_assert_eq!(552, mem::size_of::<ThreadMemberUpdate>());
+    const_assert_eq!(72, mem::size_of::<ThreadMembersUpdate>());
+    const_assert_eq!(712, mem::size_of::<ThreadUpdate>());
+    const_assert_eq!(320, mem::size_of::<TypingStart>());
+    const_assert_eq!(8, mem::size_of::<UnavailableGuild>());
+    const_assert_eq!(168, mem::size_of::<UserUpdate>());
+    const_assert_eq!(64, mem::size_of::<VoiceServerUpdate>());
+    const_assert_eq!(384, mem::size_of::<VoiceStateUpdate>());
+    const_assert_eq!(16, mem::size_of::<WebhooksUpdate>());
```

Comment:
These feel like a burden to update every time a new field is added. Is it not enough to test the ceiling of the Event size?

Reply:
Knowing the variant sizes is necessary for deciding on the optimal `Event` size. Asserting it instantly informs when a `Box` should be added / removed, instead of relying on someone periodically checking everything.

Reply:
I understand the argument. How about we go about this another way? We could test whether an event should be boxed or not by comparing it to the size of the Event. Here's what this could look like:

```rust
const EVENT_THRESHOLD: usize = 192;

const_assert!(mem::size_of::<Event>() == EVENT_THRESHOLD);

// Boxed events.
const_assert!(mem::size_of::<VoiceStateUpdate>() > EVENT_THRESHOLD);

// Unboxed events.
const_assert!(mem::size_of::<WebhooksUpdate>() <= EVENT_THRESHOLD);
```
We should be able to extract these to const functions to avoid repeating logic. It would also be nice to document instructions on what to do once one of these conditions fails compilation for those unaware.

Reply:
Unable to use `const_assert` is a const function due to [#E0401](https://doc.rust-lang.org/error-index.html#E0401) (https://github.com/nvzqz/static-assertions-rs/issues/40).